### PR TITLE
Wv 2749 locate ddv imagery

### PR DESCRIPTION
**Pull request by PatchesMaps**
## Description

>We are looking to implement functionality that allows a user to find the closest date to their currently selected date that contains imagery over the visible extent of the map. This should probably be in the form of a button that is accessible through the layer settings and gives a user a list of dates to select. Then selecting a date should update the map.

## How To Test

1. `git clone WV-2749-locate-DDV-imagery`
2. `npm ci`
3. `npm run build`
4. `npm run watch`
5. Go to this [link](http://localhost:3000/?v=-77.78111267635022,38.457486389721446,-76.36217374112248,39.27452588078139&l=Reference_Labels_15m(hidden),Reference_Features_15m(hidden),Coastlines_15m,HLS_Customizable_Landsat(bands=B07;B05;B04),HLS_Customizable_Sentinel(bands=B07;B05;B04),HLS_Shortwave_Infrared_Landsat(hidden,bands=B07;B05;B04),HLS_False_Color_Urban_Landsat(hidden,bands=B07;B06;B04),HLS_False_Color_Vegetation_Landsat(hidden,bands=B06;B05;B04),HLS_False_Color_Landsat(hidden,bands=B05;B04;B03),HLS_True_Color_Landsat(hidden,bands=B04;B03;B02),HLS_Shortwave_Infrared_Sentinel(hidden,bands=B12;B8A;B04),HLS_False_Color_Vegetation_Sentinel(hidden,bands=B11;B8A;B04),HLS_False_Color_Urban_Sentinel(hidden,bands=B12;B11;B04),HLS_False_Color_Sentinel(hidden,bands=B08;B04;B03),HLS_True_Color_Sentinel(hidden,bands=B04;B03;B02),VIIRS_NOAA20_CorrectedReflectance_TrueColor(hidden),VIIRS_SNPP_CorrectedReflectance_TrueColor(hidden),MODIS_Aqua_CorrectedReflectance_TrueColor(hidden),MODIS_Terra_CorrectedReflectance_TrueColor&lg=true&t=2023-06-02-T00%3A00%3A00Z)
6. Open the layer settings for one of the DDV layers and click "Search for Imagery".
7. Then click on the "Select Date" dropdown and select one of the populated dates to jump to a date with imagery available.

### CODE REVIEW

**Review by Tomcariello — APPROVED** [review]
This is a nice feature! I have some UI suggestions, feel free to accept or reject them:

- The dropdown might look better if it is located below the button with a matching width
- The order of the dates in the dropdown seems arbitrary. Can we sort this list ascending/descending?
- Consider limiting the user to 1 SEARCH per component since the data shouldn't change during the component lifetime.

**Review by ryanweiler92 — APPROVED** [review]

**Comment by Tomcariello** [comment]
This maybe out of scope of the ticket (the image location appears to work well), but...

If I [isolate HLS Customizable Sentinel 2A & 2B *BETA*](http://localhost:3000/?v=-77.78111267635022,38.35313664059079,-76.36217374112248,39.378875629912045&l=Reference_Labels_15m(hidden),Reference_Features_15m(hidden),Coastlines_15m(hidden),HLS_Customizable_Sentinel(bandCombo=%7B%22r%22%3A%22B07%22;%22g%22%3A%22B05%22;%22b%22%3A%22B04%22%7D),VIIRS_NOAA20_CorrectedReflectance_TrueColor(hidden),VIIRS_SNPP_CorrectedReflectance_TrueColor(hidden),MODIS_Aqua_CorrectedReflectance_TrueColor(hidden),MODIS_Terra_CorrectedReflectance_TrueColor(hidden)&lg=true&t=2023-07-12-T04%3A00%3A00Z) I see imagery on the map but the layer panel indicates that there is no imagery on any layer. Can we address that disconnect? I would expect the entry in the layer list to be active if imagery is displayed on the map.

image

Also, is there anything we can do about the URL parameters? I don't think that is a major concern but it gets a bit dicey with the band combinations i.e. "%7B%22r%22%3A%22B07%22;%22g%22%3A%22B05%22;%22b%22%3A%22B04%22%7D"

**Comment by PatchesMaps** [comment]
@Tomcariello I tweaked the CMR granule requests to cover this edgecase.

**Comment by Tomcariello** [comment]
> @Tomcariello I tweaked the CMR granule requests to cover this edgecase.
> 
> Unfortunately I think the URL parameters are out of scope for this story. Is your concern that they're too long?

I understand about scope. I'm just not used to seeing WV links with you much "%22" in them. The URL encoding just makes them long & hard to parse.

**Comment by Tomcariello** [comment]
> @Tomcariello I tweaked the CMR granule requests to cover this edgecase.

I'm still seeing the same thing. I haven't seen the layer list indicate that the HLS Customizable Sentinel 2A & 2B *BETA* layer is active in any circumstance even though I see imagery. Am I just unlucky?

**Comment by PatchesMaps** [comment]
> I'm still seeing the same thing. I haven't seen the layer list indicate that the HLS Customizable Sentinel 2A & 2B BETA layer is active in any circumstance even though I see imagery. Am I just unlucky?

Screenshot 2023-11-21 at 4 26 06 PM

It's possible that all of your CMR requests are just failing... If you open layer settings do the available dates load?

**Comment by Tomcariello** [comment]
> It's possible that all of your CMR requests are just failing... If you open layer settings do the available dates load?

If I go into the info panel I do see a number of Available Imagery Dates. Clicking on any of these dates changes the date in WV & loads new imagery but the layer stays grayed out with the blue ! zot & the "Data is not available at this location or date" alert is on the top right. It seems like all is working properly but the UI tells me there is a problem anyway.

**Comment by ryanweiler92** [comment]
I was not able to recreate the issue from Tom's link, but I am able to recreate it if I use the link in the PR description [here](http://localhost:3000/?v=-77.78111267635022,38.457486389721446,-76.36217374112248,39.27452588078139&l=Reference_Labels_15m(hidden),Reference_Features_15m(hidden),Coastlines_15m,HLS_Customizable_Landsat(bands=B07;B05;B04),HLS_Customizable_Sentinel(bands=B07;B05;B04),HLS_Shortwave_Infrared_Landsat(hidden,bands=B07;B05;B04),HLS_False_Color_Urban_Landsat(hidden,bands=B07;B06;B04),HLS_False_Color_Vegetation_Landsat(hidden,bands=B06;B05;B04),HLS_False_Color_Landsat(hidden,bands=B05;B04;B03),HLS_True_Color_Landsat(hidden,bands=B04;B03;B02),HLS_Shortwave_Infrared_Sentinel(hidden,bands=B12;B8A;B04),HLS_False_Color_Vegetation_Sentinel(hidden,bands=B11;B8A;B04),HLS_False_Color_Urban_Sentinel(hidden,bands=B12;B11;B04),HLS_False_Color_Sentinel(hidden,bands=B08;B04;B03),HLS_True_Color_Sentinel(hidden,bands=B04;B03;B02),VIIRS_NOAA20_CorrectedReflectance_TrueColor(hidden),VIIRS_SNPP_CorrectedReflectance_TrueColor(hidden),MODIS_Aqua_CorrectedReflectance_TrueColor(hidden),MODIS_Terra_CorrectedReflectance_TrueColor&lg=true&t=2023-06-02-T00%3A00%3A00Z)
ddv_imagery

File: web/js/components/feature-alert/granuleAlertModal.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { Container, Row, Col } from 'reactstrap';
+
+function GranuleAlertModalBody() {
+  return (
+    <Container className="event-alert-modal-body">
+      <Row className="p-2">
+        <Col>
+          <h2>
+            Why can&apos;t I see the data for this layer?
```
Comment (minniewong) [inline web/js/components/feature-alert/granuleAlertModal.js:10]: Change to "Why can't I see imagery for this layer?"

File: web/js/containers/sidebar/layer-row.js [hunk 2]
```diff
@@ -441,6 +481,25 @@ function LayerRow (props) {
             ))}
           </div>
         )}
+        {showZoomAlert && (
+          <AlertUtil
+            id="zoom-alert"
+            isOpen
+            title="Bad Zoom Level"
+            message="Data is not available at this zoom level."
```
Comment (minniewong) [inline web/js/containers/sidebar/layer-row.js:489]: Change to "Imagery is not available at this zoom level."

File: web/js/containers/sidebar/layer-row.js [hunk 3]
```diff
@@ -441,6 +481,25 @@ function LayerRow (props) {
...
+            onDismiss={() => setShowZoomAlert(false)}
+          />
+        )}
+        {showGranuleAlert && (
+          <AlertUtil
+            id="granule-alert"
+            isOpen
+            title="No Data Available"
```
Comment (minniewong) [inline web/js/containers/sidebar/layer-row.js:497]: "No Imagery Available"

File: web/js/containers/sidebar/layer-row.js [hunk 4]
```diff
@@ -441,6 +481,25 @@ function LayerRow (props) {
...
+            message="Data is not available at this location or date."
```
Comment (minniewong) [inline web/js/containers/sidebar/layer-row.js:498]: "Imagery not available at this location or date."

File: web/js/containers/sidebar/zot.js [hunk 5]
```diff
@@ -9,10 +9,21 @@ export default function Zot (props) {
...
+    } = zot;
     if (overZoomValue) {
       className = 'zot overzoom';
       tooltipString += `Layer is overzoomed by ${zot.overZoomValue.toString()}x its maximum zoom level. <br/>`;
+    } else if (underZoomValue) {
+      className = 'zot underzoom';
+      tooltipString += `Layer is underzoomed by ${zot.underZoomValue.toString()}x its minimum zoom level. <br/>`;
+    } else if (!hasGranules) {
+      className = 'zot no-granules';
+      tooltipString += 'No data visible data this location and/or for the date selected. <br/> Try moving the map or select a different date in the layer\'s settings <br/>';
```
Comment (minniewong) [inline web/js/containers/sidebar/zot.js:26]: "No visible imagery on this date and/or location."

File: web/js/modules/alerts/constants.js [hunk 6]
```diff
@@ -41,4 +42,15 @@ export const MODAL_PROPERTIES = {
       desktopOnly: true,
     },
   },
+  granuleModalProps: {
+    id: 'granule_layer_info',
+    props: {
+      headerText: 'Layer data is not visible at this location or date.',
```
Comment (minniewong) [inline web/js/modules/alerts/constants.js:48]: "Layer imagery is not visible at this location or date."

**Review by minniewong — CHANGES_REQUESTED** [review]
Please update the text as commented.

**Review by Tomcariello — APPROVED** [review]

**Review by rboller — CHANGES_REQUESTED** [review]
Great to see this in action!  I've sent requested changes via Slack.

**Review by ryanweiler92 — CHANGES_REQUESTED** [review]
All of the logic seems to work very well, I didn't find anything wrong there. There are just a few file structure changes I'd like to clean up. In [components/layer/settings/imagery-search ](https://github.com/nasa-gibs/worldview/blob/WV-2749-locate-DDV-imagery/web/js/components/layer/settings/imagery-search/style.css) you seem to be importing the imagery-search component into an index.js file and then exporting it again. You can probably just delete that index.js file and import your component directly from imagery-search.js. 

Also there seems to be a css style sheet within the imagery-search directory. Lets keep a consistent style format and move this to an appropriate SCSS file in the scss directory. If you don't think an appropriate file already exists for this component you can create a new one and add it to the index.scss file.
